### ISSUE TITLE
Backslashes Removed in Snippet Expansion (Potential Editor-Level Issue)

### ISSUE THREAD

**Issue by ivansigmund** [zed-industries/zed#14721]
### Check for existing issues

- [ ] Completed

### Describe the bug / provide steps to reproduce it

# Bug Report: Backslashes Removed in Snippet Expansion (Potential Editor-Level Issue)

## Description

When using custom snippets in Zed, backslashes are being removed during snippet expansion. This issue appears to affect LaTeX snippets but may be a more fundamental problem at the editor level.

## Steps to Reproduce

1. Open Zed editor
2. Add the following snippet to your Zed configuration:

```json
{
  "equation": {
    "prefix": "BEQ",
    "body": "\\begin{equation}\n\t$0\n\\end{equation}",
    "description": "equation environment"
  }
}
```

3. In a new file, type "BEQ" and trigger snippet expansion

## Expected Behavior

Typing "BEQ" should expand to:

```
\begin{equation}

\end{equation}
```

## Actual Behavior

Typing "BEQ" expands to:

```
begin{equation}

end{equation}
```

The backslashes are missing in the expanded snippet.

## Additional Information

- similar issue was also reported for the LaTeX extension lsp, where backslashes would be missing upon triggering completion : https://github.com/rzukic/zed-latex/issues/5#issue-2413244695

## Possible Cause

The backslashes might be getting escaped or removed during the snippet processing pipeline within the core editor functionality.

## Impact

This issue significantly affects the usability of snippets, especially for languages and markup formats that rely heavily on backslashes, such as LaTeX. It requires users to manually add backslashes after snippet expansion, defeating the purpose of using snippets for efficiency.

## Suggested Fix

Investigate the snippet expansion process in the core editor to ensure that backslashes are properly preserved during expansion. This may involve:

1. Reviewing the snippet parsing and expansion logic
2. Checking for any unintended escaping or string manipulation that could remove backslashes
3. Ensuring that the snippet body is treated as a literal string throughout the expansion process

### Environment

Zed: v0.144.3 (Zed)
OS: macOS 14.5.0
Memory: 8 GiB
Architecture: aarch64

### If applicable, add mockups / screenshots to help explain present your vision of the feature

https://github.com/user-attachments/assets/8fe4debd-ccbf-4bdb-9ff4-882508f3e393

### If applicable, attach your Zed.log file to this issue.

<details><summary>Zed.log</summary><pre>

</pre></details>

**Comment by lnay** [comment]
Possible [culprit](https://github.com/zed-industries/zed/blob/a072caab0b22b102f2be73189b1b5cd14eee5ae6/crates/snippet/src/snippet.rs#L51) 

Relevant example of escaping from the same commit:
```rust
    #[test]
    fn test_snippet_parsing_with_escaped_dollar_sign() {
        let snippet = Snippet::parse("\"\\$schema\": $1").unwrap();
        assert_eq!(snippet.text, "\"$schema\": ");
        assert_eq!(tabstops(&snippet), &[vec![11..11]]);
    }
```
@maxbrunsfeld from your understanding of how escaping should behave, is there any issue about accepting the "\\" in the issue as an escaped backslash?

For context, the [texlab LSP](https://github.com/latex-lsp/texlab) has been providing snippets this way.

**Comment by maxbrunsfeld** [comment]
Ok, we're probably parsing snippets wrong. I guess a backslash is only an escape character if it is followed by a `$`, otherwise it's just a literal character.

**Comment by lnay** [comment]
Thanks 🙏

**Comment by osiewicz** [comment]
This should be fixed after @lnay's fix in #14721. Give it a go.

**Comment by JosephTLyons** [comment]
This should be included in Zed [v0.146.0-pre](https://zed.dev/releases/preview/0.146.0).

**Comment by ivansigmund** [comment]
Thank you all for your work to make it work!
